protocols/purple/purple.c

```diff
@@ -1174,13 +1174,67 @@ void prplcb_conv_del_users(PurpleConversation *conv, GList *cbuddies)
 	}
 }
 
+/* Store images and replace <img> tags with urls pointing to them */
+static gboolean
+replace_images_cb (const GMatchInfo *info, GString *res, gpointer data)
+{
+	gchar *id;
+	const gchar *ext;
+	guchar *imgdata;
+	size_t imglen;
+	gchar *hash;
+	gchar *path;
+
+	if (global.conf->web_directory == NULL || global.conf->web_url == NULL) {
+		goto error;
+	}
+
+	id = g_match_info_fetch(info, 1);
+	PurpleStoredImage *image = purple_imgstore_find_by_id(atoi(id));
+
+	if (!image) {
+		goto error;
+	}
+
+	ext = purple_imgstore_get_extension(image);
+	imgdata = (guchar *) purple_imgstore_get_data(image);
+	imglen = purple_imgstore_get_size(image);
+
+	hash = g_compute_checksum_for_data(G_CHECKSUM_SHA256, imgdata, imglen);
+	path = g_strdup_printf("%s/%s.%s", global.conf->web_directory, hash, ext);
+
+	// only save images that don't exist in the filesystem, in protocols like telegram lots of images are duplicated(stickers)
+	if (access(path, F_OK) == -1) {
```

Comment:
Oh, the time-of-check-time-of-use vuln. That's not about the system call itself, it's about how it's used (`g_file_test()` can be used incorrectly like that too, [the docs mention it][1]). We're not checking for permissions here, just for existence, so the risk is that a file with the same name is created between the check and our overwrite. Non-issue imo.

[1]: https://developer.gnome.org/glib/stable/glib-File-Utilities.html#g-file-test